### ISSUE TITLE
Writing output to CSV instead of JSON

### ISSUE THREAD

**Issue by AdamDimech** [danforthcenter/plantcv#1125]
**Describe the bug**

It is not possible to output PlantCV data to a CSV file instead of a JSON file, despite [the documentation](https://plantcv.readthedocs.io/en/stable/outputs/#methods).

**To Reproduce**

Firstly, change pipeline config file to contain: 
`pcv.outputs.save_results(filename=args.result, outformat="CSV")`
or
`pcv.outputs.save_results(filename=args.result, outformat="csv")`

(it makes no difference to the outcome)

*Scenario 1* on Windows 10 [PlantCV 3.12.0, OpenCV 3.4.13.47, Python 3.9.13]:

1. Activate Python environment
2. `cd` to Output path
3. Enter the following command into the Command Prompt: `plantcv-workflow.py  -d "C:/Path/To/Project/Images"  -a filename  -p "C:/Data/projects/branching/code/pipeline.py"  -i "C:/Path/To/Project/output/2023/04/12" -j "C:/Path/To/Project/output/2023/04/12/test.json" -f plantbarcode,timestamp,measurementlabel,imgtype,other,cartag,frame  -t png  -T 16  -l "_"  --timestampformat "%Y-%m-%d %H-%M-%S"`
4. Press ENTER

*Scenario 2* on Windows 10 [PlantCV 3.12.0, OpenCV 3.4.13.47, Python 3.9.13]:

1. Activate Python environment
2. `cd` to Output path
3. Enter the following command into the Command Prompt: `plantcv-workflow.py  -d "C:/Path/To/Project/Images"  -a filename  -p "C:/Data/projects/branching/code/pipeline.py"  -i "C:/Path/To/Project/output/2023/04/12" -f plantbarcode,timestamp,measurementlabel,imgtype,other,cartag,frame  -t png  -T 16  -l "_"  --timestampformat "%Y-%m-%d %H-%M-%S"`
4. Press ENTER

*Scenario 3* on Windows 10 [PlantCV 3.12.0, OpenCV 3.4.13.47, Python 3.9.13]:

1. Activate Python environment
2. `cd` to Output path
3. Enter the following command into the Command Prompt: `plantcv-workflow.py  -d "C:/Path/To/Project/Images"  -a filename  -p "C:/Data/projects/branching/code/pipeline.py"  -i "C:/Path/To/Project/output/2023/04/12" -j "C:/Path/To/Project/output/2023/04/12/test.csv" -f plantbarcode,timestamp,measurementlabel,imgtype,other,cartag,frame  -t png  -T 16  -l "_"  --timestampformat "%Y-%m-%d %H-%M-%S"`
4. Press ENTER

*Scenario 4* on Linux [PlantCV 3.12.0, OpenCV 3.4.10.25, Python 3.8.2]:

1. Activate Python environment and PlantCV modules
2. `cd` to Output path
3. Enter the following command into the Command Prompt: `
plantcv-workflow.py -d "/group/id/path/to/images/"  -a filename  -p "/group/id/path/to/pipeline.py"  -i "/group/id/path/to/images/output/"  -j "/group/id/path/to/images/output/test.json"  -f plantbarcode,timestamp,measurementlabel,imgtype,other,cartag,frame  -t png  -T 16  -l "_"  --timestampformat "%Y-%m-%d %H-%M-%S"`
4. Press ENTER

**Expected behavior**

Data written to a CSV file or two CSV files instead of a JSON file

**Actual Behaviour**

None of these options write output to a CSV. No data file is produced. Additionally, the following output is displayed:

Scenario 1:

```
Traceback (most recent call last):
  File "C:\Users\ad06\Anaconda3\envs\pcv\Scripts\plantcv-workflow.py", line 219, in <module>
    main()
  File "C:\Users\ad06\Anaconda3\envs\pcv\Scripts\plantcv-workflow.py", line 206, in main
    plantcv.parallel.process_results(job_dir=config.tmp_dir, json_file=config.json)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\plantcv\parallel\process_results.py", line 39, in process_results
    obs = json.load(results)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\json\__init__.py", line 293, in load
    return loads(fp.read(),
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\json\__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\json\decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
2023-04-12 14:49:11,506 - distributed.client - ERROR -
Traceback (most recent call last):
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\utils.py", line 761, in wrapper
    return await func(*args, **kwargs)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\client.py", line 1225, in _reconnect
    await self._ensure_connected(timeout=timeout)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\client.py", line 1255, in _ensure_connected
    comm = await connect(
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\comm\core.py", line 289, in connect
    comm = await asyncio.wait_for(
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\asyncio\tasks.py", line 466, in wait_for
    await waiter
asyncio.exceptions.CancelledError

Traceback (most recent call last):
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\utils.py", line 761, in wrapper
    return await func(*args, **kwargs)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\client.py", line 1414, in _handle_report
    await self._reconnect()
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\utils.py", line 761, in wrapper
    return await func(*args, **kwargs)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\client.py", line 1225, in _reconnect
    await self._ensure_connected(timeout=timeout)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\client.py", line 1255, in _ensure_connected
    comm = await connect(
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\comm\core.py", line 289, in connect
    comm = await asyncio.wait_for(
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\asyncio\tasks.py", line 466, in wait_for
    await waiter
asyncio.exceptions.CancelledError
```

Scenario 2:
```
usage: plantcv-workflow.py [-h] [--config CONFIG] -d DIR [-a ADAPTOR] -p WORKFLOW -j JSON -f META [-i OUTDIR] [-T CPU] [-c] [-D DATES] [-t TYPE] [-l DELIMITER]
                           [-M MATCH] [-C COPROCESS] [-s TIMESTAMPFORMAT] [-w] [-o OTHER_ARGS] [-z CLEANUP]
plantcv-workflow.py: error: the following arguments are required: -j/--json
```

Scenario 3:

```
 File "C:\Users\ad06\Anaconda3\envs\pcv\Scripts\plantcv-workflow.py", line 219, in <module>
    main()
  File "C:\Users\ad06\Anaconda3\envs\pcv\Scripts\plantcv-workflow.py", line 206, in main
    plantcv.parallel.process_results(job_dir=config.tmp_dir, json_file=config.json)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\plantcv\parallel\process_results.py", line 39, in process_results
    obs = json.load(results)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\json\__init__.py", line 293, in load
    return loads(fp.read(),
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\json\__init__.py", line 346, in loads
    return _default_decoder.decode(s)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\json\decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\json\decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
2023-04-12 14:55:28,234 - distributed.client - ERROR -
Traceback (most recent call last):
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\utils.py", line 761, in wrapper
    return await func(*args, **kwargs)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\client.py", line 1225, in _reconnect
    await self._ensure_connected(timeout=timeout)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\client.py", line 1255, in _ensure_connected
    comm = await connect(
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\comm\core.py", line 289, in connect
    comm = await asyncio.wait_for(
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\asyncio\tasks.py", line 466, in wait_for
    await waiter
asyncio.exceptions.CancelledError

Traceback (most recent call last):
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\utils.py", line 761, in wrapper
    return await func(*args, **kwargs)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\client.py", line 1414, in _handle_report
    await self._reconnect()
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\utils.py", line 761, in wrapper
    return await func(*args, **kwargs)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\client.py", line 1225, in _reconnect
    await self._ensure_connected(timeout=timeout)
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\client.py", line 1255, in _ensure_connected
    comm = await connect(
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\site-packages\distributed\comm\core.py", line 289, in connect
    comm = await asyncio.wait_for(
  File "C:\Users\ad06\Anaconda3\envs\pcv\lib\asyncio\tasks.py", line 466, in wait_for
    await waiter
asyncio.exceptions.CancelledError
```

Scenario 4:

```
Traceback (most recent call last):
  File "/usr/local/eb/software/skylake/plantcv/3.12.0-foss-2020a-Python-3.8.2/bin/plantcv-workflow.py", line 219, in <module>
    main()
  File "/usr/local/eb/software/skylake/plantcv/3.12.0-foss-2020a-Python-3.8.2/bin/plantcv-workflow.py", line 206, in main
    plantcv.parallel.process_results(job_dir=config.tmp_dir, json_file=config.json)
  File "/usr/local/eb/software/skylake/plantcv/3.12.0-foss-2020a-Python-3.8.2/lib/python3.8/site-packages/plantcv/parallel/process_results.py", line 39, in process_results
    obs = json.load(results)
  File "/usr/local/eb/software/skylake/Python/3.8.2-GCCcore-9.3.0/lib/python3.8/json/__init__.py", line 293, in load
    return loads(fp.read(),
  File "/usr/local/eb/software/skylake/Python/3.8.2-GCCcore-9.3.0/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/local/eb/software/skylake/Python/3.8.2-GCCcore-9.3.0/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/local/eb/software/skylake/Python/3.8.2-GCCcore-9.3.0/lib/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

**The Issue**

I am unsure whether I am doing something wrong, but it seems:
- There is a requirement to use the `-j` flag to enter a path to a JSON file even if the pipeline specifies that a CSV is to be used.
- I cannot see how a single CSV file separates the single- and multiple-trait data
- The program doesn't seem to be able to handle the instructions to write to a CSV

**Comment by HaleySchuhl** [comment]
From my experience, using the `json` outformat is best practice when it comes to running workflows in parallel. We've added the `csv` option for researchers that were doing custom/per-image workflows (using annotation tools for example) so they could have data extraction and plot the results in the same jupyter notebook. However, in parallel I think we rely on json's ability to concatenate, and recommend using [json2csv](https://plantcv.readthedocs.io/en/stable/pipeline_parallel/#convert-the-output-json-file-into-csv-tables) on the command line after data extraction.

**Comment by AdamDimech** [comment]
Hi @HaleySchuhl Agreed. In fact, that is my normal practice. The reason that this issue has come to my attention is that suddenly the `pcv.outputs.add_observation` function has stopped working for me on our institute's Linux cluster (but still works locally on my Windows machine).

It's baffling, but here's what I have in my workflow:

```
uniq = len(something)

pcv.outputs.add_observation(sample='default', variable='count_unique', trait='Unique number', method='Count of structures', scale='Name', datatype=str, value=uniq, label='unique')

pcv.outputs.save_results(filename=args.result, outformat="JSON")
```

Using `print` functions inserted before and after `pcv.outputs.add_observation` I am certain that the value for `uniq` is not lost, but when the JSON file is generated (and `json2csv` implemented), the `count_unique` column is filled entirely with NAs. No errors are generated. I cannot work out where they have come from or why this is suddenly happening.

I thought I'd try writing directly to CSV to see whether there was an issue writing to JSON specifically or whether something else is going on.

I note from other conversations that a `pcv.outputs.add_metadata` was proposed but nothing came of it.

Anyway, the real issue I am trying to resolve is the sudden failure in `pcv.outputs.add_observation`. Any ideas? I am happy to send my code privately if that would help.

**Comment by HaleySchuhl** [comment]
Hmm good question. Regarding `json2csv` first, I know that there were some updates merged with PR #1099 that aimed to improve the command line tools for Windows systems. Updating your version from source would pull down these updates, since v3.14.3 was released just prior to those updates. 

Based on your sample code, the only red flag I can see why `add_observation` method would be failing is that you're providing an integer data value but telling the method that you're passing it a `str` so maybe that's mishandling the (integer?) data and making it unreadable somehow.

**Comment by nfahlgren** [comment]
PR #1099 was merged into the `release-4.0` branch, which isn't fully functional so I wouldn't use that quite yet.

@AdamDimech if you run the workflow on a single image and you look at the output JSON file (`args.result`), is your custom observation in that file? It sounded like it wasn't working in parallel but want to make sure it's also not working for a single run
